### ISSUE TITLE
Aquamacs 3.4 (and 3.5) chokes on "unsafe" file local variables while eval'ing init file

### ISSUE THREAD

**Issue by nega0** [aquamacs-emacs/aquamacs-emacs#147]
When starting Aquamacs with `aquamacs`, `open -a Aquamacs.app` or `/Applications/Aquamacs.app/Contents/MacOS/Aquamacs` Aquamacs will "hang" without drawing a window. The macOS menubar will display " Aquamacs" then the "full" Aquamacs menu. Clicking "Windows" will give:

![screen shot 2019-02-12 at 10 11 34 am](https://user-images.githubusercontent.com/240504/52647074-0a195700-2eb2-11e9-921a-6bd0ec559abd.png)

A "Force Quit" is then required to close Aquamacs. I assume that the "•Local Variables•" buffer is the dialog where "Emacs" normally asks you to confirm the safety of unknown file local variables. As stated in the manual Emacs shouldn't be prompting the user in "batch" mode, and instead assuming 'n'.

Removing the offending file local variable from `~/.emacs` restores expected behavior; Aquamacs starts up properly.

tl;dr
Aquamacs choking on unknown/unsafe file local variable while eval'ing init file. This behavior is contrary to Aquamacs 3.3, and Emacs.

**Comment by treese** [comment]
I have had this problem a lot, mostly with using `desktop` to restore open files.

The proximate problem seems to be around line 1088 in `lisp/startup.el`, which suppresses the visibility of the initial frame. The commit for that is just for Aquamacs, with a commit message about dealing with a problem with `--debug-init`. I don't know what that problem was, so I hesitate to simply take it out.

However, I did rebuild Aquamacs with this code omitted:

     (let ((initial-frame-alist (append (unless init-file-debug '((visibility . nil)))
				      initial-frame-alist)))
     ...

(by "omitted" I mean properly redone to deal with a missing let, not just deleting those lines, of course!)

For a simple test using the file `/tmp/localtest.el` with the contents:

    ;; Local variables:
    ;; test-var: nil
    ;; end:

and the command `open -a Aquamacs --args -q /tmp/localtest.el`. 

Aquamacs standard release and nightly build both hang as described. With the code omitted, the startup proceeds and displays a buffer prompting about the local variables as expected.

**Comment by davidswelt** [comment]
The code is there so that there is not an ugly cacophony of flashing things and opening and closing windows during the startup sequence. A better solution to this issue -- and I am surprised it hasn't come up until now -- is to make sure the frame gains visibility the moment interaction is required.  One way to do that may be `minibuffer-setup-hook'.

**Comment by treese** [comment]
`minibuffer-setup-hook` doesn't quite seem to do it, because the local variables prompt doesn't use the minibuffer; it pops up a regular buffer. With the current Aquamacs code, do you know where in the startup sequence the frame becomes visible? The standard startup is described at https://www.gnu.org/software/emacs/manual/html_node/elisp/Startup-Summary.html. Maybe from there we could work backward to a way to solve this.

**Comment by davidswelt** [comment]
On Mar 29, 2019, at 10:03 PM, Win Treese <notifications@github.com> wrote:
> 
> minibuffer-setup-hook doesn't quite seem to do it, because the local variables prompt doesn't use the minibuffer; it pops up a regular buffer. With the current Aquamacs code, do you know where in the startup sequence the frame becomes visible? The standard startup is described at https://www.gnu.org/software/emacs/manual/html_node/elisp/Startup-Summary.html. Maybe from there we could work backward to a way to solve this.

Here:

      ;; time to make the frame visible (Aquamacs)
      (unless (and (eq initial-window-system 'ns)
		   (ns-application-hidden-p))
	(if (or show-scratch-buffer-on-startup
		(not (equal (buffer-name (current-buffer)) "*scratch*")))
	    (make-frame-visible)))


The most pragmatic way would be to (make-frame-visible) in `hack-local-variables-confirm’.


--
http://aquamacs.org -- Aquamacs: Emacs on Mac OS X
http://aquamacs.org/donate -- Could we help you? Return the favor and support the Aquamacs Project!

**Comment by treese** [comment]
I can make the change in `hack-local-variables-confirm` and try it out if you want to patch it that way.

**Comment by davidswelt** [comment]
Sounds good

**Comment by davidswelt** [comment]
Win, has this been changed?

**Comment by treese** [comment]
I'll take a look at it.

**Comment by treese** [comment]
Your idea seemed to work, so I committed the one-line change directly into the current code. It would be good to observe this from the nightly for a few days to see if there are any unintended side effects.

**Comment by nega0** [comment]
With the August 02 nightly, the expected behavior seems to be restored. Using my daily config Aquamacs starts as usual. With a fresh config, Aquamacs pops up the frame with the confirmation dialog as expected.

_worksforme_

Thank you!

**Comment by nega0** [comment]
At work, with my daily config, the confirmation menu pops up, but it pops up every time Aquamacs starts (yes ever after hitting `!` to permanently mark those variables as safe).

**Comment by treese** [comment]
Can you give an example of a file that triggers the problem with without other configuration?

**Comment by nega0** [comment]
I've narrowed it down a bit. 8941ba0fd04d84f783e067b6b67380b7bc7a5067 works provided you're using `~/.emacs.d/init.el` or `~/.emacs.el`. If your init file is `~/.emacs` (file or symlink), the safe-local-variables-dialog seems to record your answer (`!`), but doesn't read it on the next startup.

> Can you give an example of a file that triggers the problem with without other configuration?

Just pick any unsafe file local variable.
`$ echo ';; -*- fake-local-variable: poop -*-' > ~/.emacs`

**Comment by treese** [comment]
@nega0 I can reproduce a problem with this, though it's not exactly the same. Do you use desktop-save-mode, revive, or anything else that restores files on startup?

**Comment by nega0** [comment]
No, nothing that restores files on startup. The closest things would be `saveplace`, `recentf`, and `smart-frame-positioning`, as a result of Aquamacs defaults.

**Comment by treese** [comment]
Lost track of this one. Will try to fix in 3.6.

**Comment by treese** [comment]
OK, this is a little twisty. I don't fully understand the Aquamacs initialization process, but here are some clues:

In `startup.el` (a modified standard Emacs file), the code that loads init files does the following:

1. Tries to load `.emacs` or an OS-dependent variation of that name. There is no error here if the file is not found.
2. This happens before the custom file that contains the safe local variables is loaded, which is why doesn't appear to have been saved.
3. Aquamacs loads custom and preferences files
4. If alternatives to `.emacs` are used, like `.emacs.el` are used, those get run after the custom and Preferences files, so the safe local variables are available.

I think the best option is to leave it the way it is, because it's a tiny corner case that's unlikely for Aquamacs users (that is, the combination of a file local variable for specifically `.emacs` configurations), and I'm concerned that changing any of the order of initialization would had some bad side effects for other users.

For the moment, I'm going to leave this open for your comments, and I'm changing the milestone because it won't be changed for 3.6.

Thoughts welcome!

**Comment by nega0** [comment]
Based on your last update, I don't think any further action is necessary. I've migrated to using `~/.emacs.d/init.el`, which is what I believe the Emacs manual suggests anyway. Of course Aquamacs users should use the Aquamacs way :) We can consider this "fixed via documentation".
